Update `--require`

tommy-mitchell opened the pull request:
Implements and closes #4 using a slightly modified `import-or-require.js` from [`tape`](https://github.com/substack/tape/blob/v5.6.0/bin/import-or-require.js). Adds [`get-package-type`](https://www.npmjs.com/package/get-package-type) as a dependency[^1].

Modifies the `cli()` function in `cli.js` for module resolution:

```js
for (const module of args.require) {
    const resolved = nodeResolve(module, {basedir: cwd});
    importOrRequire(resolved);
}

// ...

for (const file of files) {
    // always resolve before import for windows
    const resolved = resolvePath(cwd, file);
    promises.push(importOrRequire(resolved));
}
```

Currently:

https://github.com/coderaiser/supertape/blob/f37329cc32121d268dd259e6bfccb32967422527/packages/supertape/lib/cli.js#L145-L146

https://github.com/coderaiser/supertape/blob/f37329cc32121d268dd259e6bfccb32967422527/packages/supertape/lib/cli.js#L179-L183

----

This PR does not remove `simple-import.js`.

[^1]: `tape` currently has `resolve` set to [`^2.0.0-next.3`](https://github.com/browserify/resolve/tree/v2.0.0-next.3) (vs. `^1.17.0` in Supertape). Haven't looked into the changes between the two versions, but `--require` seems to function correctly with both.

coderaiser commented:
Maybe we should add one more flag `—-import` that will do the thing, what do you think? It will simplify code (and tests) a lot.

coderaiser commented on packages/supertape/lib/cli.js:
```diff
@@ -178,8 +180,8 @@ async function cli({argv, cwd, stdout, isStop}) {
-        const resolved = pathToFileURL(resolvePath(cwd, file));
-        promises.push(simpleImport(resolved));
+        const resolved = resolvePath(cwd, file);
+        promises.push(importOrRequire(resolved));
```
So you want to add it to every loaded file so it was `imported` or `required`? Why not just use `ts-node` for this purpose 🤔?

tommy-mitchell replied:
A better solution that simply checks the extension:

```js
if (extnamePath(file) === '.ts') {
  const resolved = resolvePath(cwd, file);
  promises.push(require(resolved));
} else {
  const resolved = pathToFileURL(resolvePath(cwd, file));
  promises.push(simpleImport(resolved));
}
```

Maybe it would be better to check that the extension does not contain `js`? In case somebody wants to use this with another superset language of JavaScript.

tommy-mitchell commented:
Can probably revert to this as well:

https://github.com/coderaiser/supertape/blob/f37329cc32121d268dd259e6bfccb32967422527/packages/supertape/lib/cli.js#L145-L146

I had an issue where using a symlinked install of Supertape causes this to fail (necessitating `resolve`), however using a regular `npm install supertape` this works fine. I can run CJS, ES, and TS tests all in the same `supertape test.*` command (if `--require`ing the transpiler, of course)

tommy-mitchell commented:
I've simplified this change down and removed the added dependency. The current implementation now just `require()`s non-JS files instead of `import()`ing them:

```js
// cli.js, in function cli()

// now the same as in master
for (const module of args.require)
    importOrRequire(module);

// ...

for (const file of files) {
  // always resolve before import for windows
  if (extnamePath(file).includes('js')) {
    const resolved = pathToFileURL(resolvePath(cwd, file));
    promises.push(simpleImport(resolved));
  } else {
    const resolved = resolvePath(cwd, file);
    promises.push(require(resolved));
  }
}
```

tommy-mitchell committed:
Move CLI file handling to `import-or-require.js`

tommy-mitchell commented:
I've made this change far simpler. I've re-added `import-or-require.js` to handle file resolution:

```js
// cli.js

for (const file of files)
    promises.push(importOrRequire(cwd, file));
```

Now, `import-or-require.js` handles resolving the path to the given file. If said file has an extension ending in `js` (e.g. `.js`, `.cjs`, `.mjs`), then nothing new happens (same as `simple-import.js`). If, however, the extension is different, then the file is `require`ed instead:

```js
// import-or-require.js

module.exports.importOrRequire = (cwd, file) => {
    const resolved = resolvePath(cwd, file);
    
    if (extnamePath(file).endsWith('js')) {
        // always resolve before import for windows
        import(pathToFileURL(resolved));
    } else {
        require(resolved);
    }
};
```

If I have a set of tests `test.js`, `test.cjs`, `test.mjs`, and `test.ts`, I can run all of them with:

```
$ supertape -r ts-node/register/transpile-only test.*
```

`simple-import.js` has not been deleted, but is not currently being used.

coderaiser commented:
> Now, import-or-require.js handles resolving the path to the given file. If said file has an extension ending in js (e.g. .js, .cjs, .mjs), then nothing new happens (same as simple-import.js). If, however, the extension is different, then the file is requireed instead

In this case ESM loaders will not handle the case with ts or jsx.

Have you tried using ESM loader for ts?

tommy-mitchell commented:
> In this case ESM loaders will not handle the case with ts or jsx.

Whoops, didn't consider `.jsx`. Will change it to `EXT.includes('js')`.

> Have you tried using ESM loader for ts?

Hm, I hadn't considered this. Seems like `tape` is [running into this issue](https://github.com/substack/tape/issues/577) as well. I'm not sure what the solution is - I did a quick search and couldn't seem to find a package to help. Perhaps a dirty solution is to attempt to `require()` the file, and if `node` throws a `SyntaxError: Cannot use import statement outside a module`, then we catch the error and `import()` the file instead.

A workaround for users could be to specify that `ts-node` emits CJS modules for tests:

```jsonc
// tsconfig.json, but could be in package.json as well
{
  "compilerOptions": {
    "module": "ES6",
  },
  "ts-node": {
    // could have "include" or "files" here to limit to tests
    "compilerOptions": {
      "module": "CommonJS"
    }
  }
}
```

tommy-mitchell committed:
Fix `js` extension detection

tommy-mitchell commented:
How do I write a test for `cli.spec.js`? I've added some fixtures with different extensions:

```
lib/
\__ fixture/
    \__ import-or-require/
        \__ test.js
        \__ test.cjs
        \__ test.mjs
        \__ test.jsx
        \__ test.ts
\__ cli.js
\__ cli.spec.js
\__ supertape.js
```

Each one runs 50 tests, with the import statements and test message changing respectively per file/iteration:

```js
// test.js

const test = require('../../supertape');

for(const i in [...Array(50).keys()]) {
    test(`import or require: .js ${i}`, async (t) => {
        await new Promise(r => setTimeout(r, 5));

        t.pass('import or require with supertape binary');
        t.end();
    });
}
```

I can run all of them by requiring the appropriate transpilers:

```sh
$ supertape -r 'ts-node/register' -r '@babel/register' 'fixture/import-or-require/test.*'
```

```
TAP version 13

1..250
# tests 250
# pass 250

# ✅ ok
```

However, I'm struggling to write a test that does the same thing as the `supertape` binary command. I know it's something like this:

```js
// cli.spec.js

test('supertape: cli: -r: import or require', async (t) => {
    const argv = [
        '-r', 'ts-node/register',
        '-r', '@babel/register',
        'fixture/import-or-require/test.*',
    ];

    const [error, cli] = await runCli({
        argv,
    });

    // assertion with CLI result
    t.end();
});
```

But whenever I try to run this test, none of the files are included.
